subsys/bluetooth/host/audio/broadcast_source.c

```diff
@@ -58,11 +58,157 @@ struct bt_audio_base_ad {
 	struct bt_audio_base_ad_subgroup subgroups[BROADCAST_SUBGROUP_CNT];
 } __packed;
 
-static struct bt_audio_broadcast_source broadcast_sources[BROADCAST_SRC_CNT];
+static struct bt_audio_ep broadcast_source_eps
+	[CONFIG_BT_AUDIO_BROADCAST_SRC_COUNT][BROADCAST_STREAM_CNT];
+static struct bt_audio_broadcast_source broadcast_sources[CONFIG_BT_AUDIO_BROADCAST_SRC_COUNT];
 
 static int bt_audio_set_base(const struct bt_audio_broadcast_source *source,
 			     struct bt_codec *codec);
 
+static void broadcast_source_set_ep_state(struct bt_audio_ep *ep, uint8_t state)
+{
+	uint8_t old_state;
+
+	old_state = ep->status.state;
+
+	BT_DBG("ep %p id 0x%02x %s -> %s", ep, ep->status.id,
+	       bt_audio_ep_state_str(old_state),
+	       bt_audio_ep_state_str(state));
+
+
+	switch (old_state) {
+	case BT_AUDIO_EP_STATE_IDLE:
+		if (state != BT_AUDIO_EP_STATE_QOS_CONFIGURED) {
+			BT_DBG("Invalid broadcast sync endpoint state transition");
+			return;
+		}
+		break;
+	case BT_AUDIO_EP_STATE_QOS_CONFIGURED:
+		if (state != BT_AUDIO_EP_STATE_IDLE &&
+		    state != BT_AUDIO_EP_STATE_STREAMING) {
+			BT_DBG("Invalid broadcast sync endpoint state transition");
+			return;
+		}
+	case BT_AUDIO_EP_STATE_STREAMING:
+		if (state != BT_AUDIO_EP_STATE_QOS_CONFIGURED) {
+			BT_DBG("Invalid broadcast sync endpoint state transition");
+			return;
+		}
+		break;
+	default:
+		BT_ERR("Invalid broadcast sync endpoint state: %s",
+		       bt_audio_ep_state_str(old_state));
+		break;
+	}
+
+	ep->status.state = state;
+
+	if (state == BT_AUDIO_EP_STATE_IDLE) {
+		struct bt_audio_stream *stream = ep->stream;
+
+		if (stream != 0) {
+			stream->ep = NULL;
+			stream->codec = NULL;
+			ep->stream = NULL;
+		}
+	}
+}
```

Comment:
This code seems 100% identical to the sink version of the function - if you have more functions that are identical you could put them into a broadcast_common.c file. When role-specific changes are needed they could be moved to the respective files again.
This is just a suggestion - there is also great value in having the code separated as it makes it easier to make changes as one only needs to consider the specific role that is changed - hence less risk of breaking things ;-)

Reply:
Agreed. This function also look fairly large, but is actually very small when debug is disabled:

```
	ep->status.state = state;

	if (state == BT_AUDIO_EP_STATE_IDLE) {
		struct bt_audio_stream *stream = ep->stream;

		if (stream != 0) {
			stream->ep = NULL;
			stream->codec = NULL;
			ep->stream = NULL;
		}
	}
```

But of course if the statemachine for broadcast source and sink will stay the same (the broadcast sink may not end up having the `BT_AUDIO_EP_STATE_QOS_CONFIGURED` state), we could have a common function for it.